candidate-agreement/src/table.rs

```diff
@@ -0,0 +1,986 @@
+			known_statements: HashSet::default(),
+		}).known_statements.insert(trace);
+	}
+
+	fn import_candidate(
+		&mut self,
+		context: &C,
+		from: C::ValidatorId,
+		candidate: C::Candidate,
+		signature: C::Signature,
+	) -> (Option<<C as ResolveMisbehavior>::Misbehavior>, Option<Summary<C::Digest, C::GroupId>>) {
+		let group = context.candidate_group(&candidate);
+		if !context.is_member_of(&from, &group) {
+			return (
+				Some(Misbehavior::UnauthorizedStatement(UnauthorizedStatement {
+					statement: SignedStatement {
+						signature,
+						statement: Statement::Candidate(candidate),
+						sender: from,
+					},
+				})),
+				None,
+			);
+		}
+
+		// check that validator hasn't already specified another candidate.
+		let digest = context.candidate_digest(&candidate);
+
+		match self.validator_data.entry(from.clone()) {
+			Entry::Occupied(mut occ) => {
+				// if digest is different, fetch candidate and
+				// note misbehavior.
+				let existing = occ.get_mut();
+
+				if let Some((ref old_digest, ref old_sig)) = existing.proposal {
+					if old_digest != &digest {
+						let old_candidate = self.candidate_votes.get(old_digest)
+							.expect("proposed digest implies existence of votes entry; qed")
+							.candidate
+							.clone();
+
+						return (
+							Some(Misbehavior::MultipleCandidates(MultipleCandidates {
+								first: (old_candidate, old_sig.clone()),
+								second: (candidate, signature.clone()),
+							})),
+							None,
+						);
+					}
+				} else {
+					existing.proposal = Some((digest.clone(), signature.clone()));
+				}
+			}
+			Entry::Vacant(vacant) => {
+				vacant.insert(ValidatorData {
+					proposal: Some((digest.clone(), signature.clone())),
+					known_statements: HashSet::new(),
+				});
+
+				// TODO: seed validity votes with issuer here?
+				self.candidate_votes.entry(digest.clone()).or_insert_with(move || CandidateData {
+					group_id: group,
+					candidate: candidate,
+					validity_votes: HashMap::new(),
+					availability_votes: HashMap::new(),
+					indicated_bad_by: Vec::new(),
+				});
+			}
+		}
+
+		self.validity_vote(
+			context,
+			from,
+			digest,
+			ValidityVote::Issued(signature),
+		)
+	}
+
+	fn validity_vote(
+		&mut self,
+		context: &C,
+		from: C::ValidatorId,
+		digest: C::Digest,
+		vote: ValidityVote<C::Signature>,
+	) -> (Option<<C as ResolveMisbehavior>::Misbehavior>, Option<Summary<C::Digest, C::GroupId>>) {
+		let votes = match self.candidate_votes.get_mut(&digest) {
+			None => return (None, None), // TODO: queue up but don't get DoS'ed
+			Some(votes) => votes,
+		};
+
+		// check that this validator actually can vote in this group.
+		if !context.is_member_of(&from, &votes.group_id) {
+			let (sig, valid) = match vote {
+				ValidityVote::Valid(s) => (s, true),
+				ValidityVote::Invalid(s) => (s, false),
+				ValidityVote::Issued(_) =>
+					panic!("implicit issuance vote only cast if the candidate entry already created successfully; qed"),
```

Comment:
could do to be a little more explicit and structured. e.g. `validity_vote` only called with `ValidityVote::Issued` by `import_candidate`; `import_candidate` callsite unreachable unless `from` is a member of the candidate's group.